pkg/build/gobuild.go

```diff
@@ -563,3 +560,72 @@ func updatePath(cf *v1.ConfigFile) {
 	// If we get here, we never saw PATH.
 	cf.Config.Env = append(cf.Config.Env, "PATH="+appDir)
 }
+
+// Build implements build.Interface
+func (gb *gobuild) Build(ctx context.Context, s string) (Result, error) {
+	// Determine the appropriate base image for this import path.
+	base, err := gb.getBase(s)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine what kind of base we have and if we should publish an image or an index.
+	mt, err := base.MediaType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch mt {
+	case types.OCIImageIndex, types.DockerManifestList:
+		base, ok := base.(v1.ImageIndex)
+		if !ok {
+			return nil, fmt.Errorf("failed to interpret base as index: %v", base)
+		}
+		return gb.buildAll(ctx, s, base)
+	case types.OCIManifestSchema1, types.DockerManifestSchema2:
+		base, ok := base.(v1.Image)
+		if !ok {
+			return nil, fmt.Errorf("failed to interpret base as image: %v", base)
+		}
+		return gb.buildOne(ctx, s, base)
+	default:
+		return nil, fmt.Errorf("base image media type: %s", mt)
+	}
+}
+
+func (gb *gobuild) buildAll(ctx context.Context, s string, base v1.ImageIndex) (v1.ImageIndex, error) {
+	im, err := base.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	// Build an image for each child from the base and append it to a new index to produce the result.
+	adds := []mutate.IndexAddendum{}
+	for _, desc := range im.Manifests {
+		// TODO(jonjohnsonjr): This would fail for a nested index, which is exceedingly rare.
+		base, err := base.Image(desc.Digest)
+		if err != nil {
+			return nil, err
+		}
+		img, err := gb.buildOne(ctx, s, base)
```

Comment:
Should we do these in parallel?

Reply:
`TODO(#1234)` would be fine, I think

Reply:
I would be worried even more about cache invalidation, will file an issue to experiment and follow-up, then TODO(it).

Reply:
Done.